TypeCobol/Compiler/Diagnostics/TypeCobolChecker.cs

```diff
@@ -43,70 +44,163 @@ class ReadOnlyPropertiesChecker: NodeListener {
+                Check(node.CodeElement, node.SymbolTable, procedureStyleCall.FunctionCall, functionDeclarations);
             }
-	}
-	private void Check(CodeElement e, SymbolTable table, [NotNull] FunctionCall call,
-	    [NotNull] FunctionDeclaration definition) {
-		var parameters = definition.Profile.Parameters;
-        var callArgsCount = call.Arguments != null ? call.Arguments.Length : 0;
-        if (callArgsCount > parameters.Count) {
-			var m = string.Format("Function {0} only takes {1} parameters", definition.Name, parameters.Count);
-			DiagnosticUtils.AddError(e, m);
-		}
-		for (int c = 0; c < parameters.Count; c++) {
-			var expected = parameters[c];
-			if (c < callArgsCount) {
-				var actual = call.Arguments[c].StorageAreaOrValue;
-				if (actual.IsLiteral) continue;
-                var callArgName = actual.MainSymbolReference != null ? actual.MainSymbolReference.Name : null;
-                var found = table.GetVariable(actual);
-				if (found.Count < 1) DiagnosticUtils.AddError(e, "Parameter "+callArgName+" is not referenced");
-				if (found.Count > 1) DiagnosticUtils.AddError(e, "Ambiguous reference to parameter "+callArgName);
-				if (found.Count!= 1) continue;
-				var type = found[0] as ITypedNode;
-				// type check. please note:
-				// 1- if only one of [actual|expected] types is null, overriden DataType.!= operator will detect it
-				// 2- if both are null, we WANT it to break: in TypeCobol EVERYTHING should be typed,
-				//    and things we cannot know their type as typed as DataType.Unknown (which is a non-null valid type).
-				if (type == null || type.DataType != expected.DataType) {
-					var m = string.Format("Function {0} expected parameter {1} of type {2} (actual: {3})", definition.Name, c+1, expected.DataType, type.DataType);
-					DiagnosticUtils.AddError(e, m);
-				}
-				if (type.Length > expected.Length) {
-					var m = string.Format("Function {0} expected parameter {1} of max length {2} (actual: {3})", definition.Name, c+1, expected.Length, type.Length);
-					DiagnosticUtils.AddError(e, m);
-				}
-			} else {
-				var m = string.Format("Function {0} is missing parameter {1} of type {2}", definition.Name, c+1, expected.DataType);
-				DiagnosticUtils.AddError(e, m);
-			}
-		}
-	}
-}
+        }
+
+        private int CheckFunctionVariableAmbiguity(List<DataDefinition> potentialVariables,
+            List<FunctionDeclaration> functionDeclarations, SymbolTable symbolTable)
+        {
+            return functionDeclarations.Where(f => potentialVariables.FindAll(v => v.Name == f.Name).Count == 1).Count();
+        }
+
+        private List<FunctionDeclaration> CheckFunctionAmbiguity(List<FunctionDeclaration> functionDeclarations,
+            Node node)
+        {
+            if (functionDeclarations.Count > 1)
+            {
+                var m = string.Format("Function {0} is ambigous",
+                    ((ProcedureStyleCallStatement) node.CodeElement).ProcedureCall.FunctionName);
```

Comment:
You can add FunctionCaller as a parameter to avoid a cast to ProcedureStyleCallStatement